### PR TITLE
config: support TLS in iproto.listen and iproto.advertise.{peer, sharding}

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch-set introduces new format for `iproto.listen` and `iproto.advertise.{peer, sharding}` and introduces support for TLS in these options.

Part of #8862

### CODE REVIEW

**Review by andreyaksenov — APPROVED** [review]

**Review by ochaplashkin — APPROVED** [review]

File: test/config-luatest/instance_config_schema_test.lua [hunk 1]
```diff
@@ -43,7 +43,7 @@ local function validate_fields(config, record)
                 validate_fields(v1, v.items)
             end
         end
-        if v.type ~= 'record' or not v.enterprise_edition or is_enterprise then
+        if not v.enterprise_edition or is_enterprise then
```
Comment (Totktonada) [inline test/config-luatest/instance_config_schema_test.lua:46]: Just for the record, it was discussed earlier: https://github.com/tarantool/tarantool/pull/9069#discussion_r1311019517.

File: src/box/lua/config/instance_config.lua [hunk 2]
```diff
@@ -653,25 +656,49 @@ return schema.new('instance_config', schema.record({
...
+                    })),
+                    ssl_cert_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_ciphers = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_key_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:692]: What is important about all the file paths in the configuration: it should point to the same file before and after the first `box.cfg()` call even if the path is relative and `process.work_dir` is set.

I don't know, whether `box.cfg()` handles it on its own and, if so, how exactly.

Please, check this point. If a special handing is required, consider generalizing code from `safe_mkdir()` (`applier/mkdir.lua`).

File: src/box/lua/config/instance_config.lua [hunk 3]
```diff
@@ -653,25 +656,49 @@ return schema.new('instance_config', schema.record({
         end,
     }),
     iproto = schema.record({
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua:642]: Renamed commit to `support TLS options in ...`. Added doc-request.

File: src/box/lua/config/instance_config.lua [hunk 4]
```diff
@@ -653,25 +656,49 @@ return schema.new('instance_config', schema.record({
...
+                        -- Substitute variables with placeholders to don't
+                        -- confuse the URI parser with the curly brackets.
+                        data = data:gsub('{{ *.- *}}', 'placeholder')
+                        local uris, err = urilib.parse_many(data)
+                        if uris == nil then
+                            w.error('Unable to parse an URI/a list of URIs: %s',
+                                    err)
+                        end
+                    end,
+                }),
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Done. However, it is still allowed to have login and password in field `uri` of element of `iproto.listen`.

File: test/config-luatest/config_test.lua [hunk 5]
```diff
@@ -921,3 +934,140 @@ g.test_failover_supervised_constrainsts = function(g)
...
+            ['group-001'] = {
+                replicasets = {
+                    ['replicaset-001'] = {
+                        instances = {
+                            ['instance-001'] = {},
+                        },
+                    },
+                },
+            },
+        },
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/config_test.lua]: Done.

File: test/config-luatest/config_test.lua [hunk 6]
```diff
@@ -921,3 +934,140 @@ g.test_failover_supervised_constrainsts = function(g)
...
+    }
+
+    treegen.write_script(dir, 'config.yaml', require('yaml').encode(config))
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/config_test.lua]: Done.

File: src/box/lua/config/instance_config.lua [hunk 7]
```diff
@@ -653,25 +656,49 @@ return schema.new('instance_config', schema.record({
...
+                        type = 'string',
+                    })),
+                    ssl_password = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                    ssl_password_file = enterprise_edition(schema.scalar({
+                        type = 'string',
+                    })),
+                }),
+            }),
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Done, I used the second option.

File: test/config-luatest/cluster_config_schema_test.lua [hunk 8]
```diff
@@ -182,7 +182,6 @@ g.test_defaults = function()
             snap_io_rate_limit = box.NULL,
         },
         iproto = {
-            listen = box.NULL,
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/cluster_config_schema_test.lua:185]: Yes. I added code that replaces `nil` to `box.NULL` in box_cfg.lua.

File: test/config-luatest/config_test.lua [hunk 9]
```diff
@@ -921,3 +934,140 @@ g.test_failover_supervised_constrainsts = function(g)
...
+    })
+end
+
+g.test_iproto_listen_ssl_enterprise = function()
+    t.tarantool.skip_if_not_enterprise()
+    local dir = treegen.prepare_directory(g, {}, {})
+    local passwd_file = fio.pathjoin(dir, 'passwd.txt')
+    local file = fio.open(passwd_file, {'O_WRONLY', 'O_CREAT'},
+                          tonumber('666', 8))
+    t.assert(file ~= nil)
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/config_test.lua:1075]: I added file:close(). Also, I still use both options here, but I added a new test, where one place uses `ssl_password` and another uses `ssl_password_file`.

File: test/config-luatest/config_test.lua [hunk 10]
```diff
@@ -921,3 +934,140 @@ g.test_failover_supervised_constrainsts = function(g)
...
+        assert(box.cfg.listen[1].params.ssl_password_file == '%s')
+        assert(box.cfg.listen[1].params.ssl_ciphers == '%s')
+        os.exit(0)
+    ]]):format(ca_file, cert_file, key_file, passwd, passwd_file, ciphers)
+    treegen.write_script(dir, 'main.lua', script)
+
+    local args = {'--name', 'instance-001', '--config', config_file, 'main.lua'}
+    local opts = {nojson = true, stderr = false}
+    local res = justrun.tarantool(dir, {}, args, opts)
+    t.assert_equals(res.exit_code, 0)
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/config_test.lua]: I've reworked the test so it uses the latest Hepler server.

File: src/box/lua/config/instance_config.lua (hunk 2 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: We checked that there were no problems and all relative `ssl_*_file` paths were found in `process.work_dir`. There is, however, no test for now.

File: test/config-luatest/config_test.lua [hunk 11]
```diff
@@ -921,3 +934,140 @@ g.test_failover_supervised_constrainsts = function(g)
         exp_err = bootstrap_strategy_exp_err_template:format('supervised'),
     })
 end
+
+g.test_iproto_listen_plain = function()
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply test/config-luatest/config_test.lua:1036]: Added.

**Review by ochaplashkin — APPROVED** [review]

File: src/box/lua/config/instance_config.lua (hunk 2 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: I'm OK to leave it as `box.cfg()` deal.

File: src/box/lua/config/instance_config.lua [hunk 12]
```diff
@@ -638,25 +640,64 @@ return schema.new('instance_config', schema.record({
...
+                        type = 'string',
+                    })),
+                }),
+            }, {
+                validate = function(data, w)
+                    -- No items in iproto.listen at all -- OK.
+                    if data == nil or next(data) == nil then
+                        return
+                    end
+                    -- There is some data -- the prefix should be there.
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:694]: Typo: prefix.

BTW, should an empty record be an error too? Here we have an array of such records and if there is no need to set the URI, one can just don't add the item.

File: src/box/lua/config/instance_config.lua [hunk 13]
```diff
@@ -736,10 +691,9 @@ return schema.new('instance_config', schema.record({
         -- The iproto.advertise.{peer,sharding} options have the
         -- following syntax variants:
         --
-        -- 1. user@
-        -- 2. user:pass@
-        -- 3. user@host:port
-        -- 4. host:port
+        -- {login = ..., password = ..., uri = ..., params = ...}
+        --
+        -- where uri is host:port.
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:737]: Nit: I would say:

> the following syntax variants:
>
> 1. {login = ...}
> 2. {login = ..., password = ...}
> 3. {uri = ...[, params = ...]}
> 4. {uri = ..., login = ...[, params = ...]}
> 5. {uri = ..., login = ..., password = ...[, params = ...]}

It shows what is allowed and what is not.

(BTW, the item numbers are used in this comment below.)

Or just

> the following syntax:
>
> {login = ..., password = ..., uri = ..., params = ...}

File: src/box/lua/config/instance_config.lua [hunk 14]
```diff
@@ -773,15 +727,152 @@ return schema.new('instance_config', schema.record({
...
+            sharding = schema.record({
+                login = schema.scalar({
+                    type = 'string',
+                }),
+                password = schema.scalar({
+                    type = 'string',
+                }),
+                uri = schema.scalar({
+                    type = 'string',
+                    validate = function(data, w)
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:853]: Nit: `iproto.advertise.{peer,sharding}` validation function looks the same. Can we reduce the code duplication?

File: src/box/lua/config/instance_config.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: > However, it is still allowed to have login and password in field `uri` of element of `iproto.listen`.

Hm. It has no sense in context of a listening socket, it constradicts with the advertise URI format (with separate login/password), it influences replication peer URI calculation (that may be surprising, when there is advertise URI for it). I think there should be a good reason to allow it, but I don't see such a reason.

File: src/box/lua/config/instance_config.lua (hunk 12 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Thanks, fixed.

File: src/box/lua/config/instance_config.lua (hunk 13 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: True, fixed.

File: src/box/lua/config/instance_config.lua (hunk 14 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Moved this validation to a function.

File: src/box/lua/config/instance_config.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Fixed. I have disabled login and password in the URI.

File: src/box/lua/config/configdata.lua [hunk 15]
```diff
@@ -91,6 +91,9 @@ local function instance_sharding(iconfig, instance_name)
     end
     local zone = instance_config:get(iconfig, 'sharding.zone')
     local uri = instance_config:instance_uri(iconfig, 'sharding')
+    if uri == nil then
+        error('No suitable URI provided', 0)
+    end
```
Comment (Totktonada) [inline src/box/lua/config/configdata.lua:96]: Nit: Shouldn't we report an instance name at least?

File: src/box/lua/config/instance_config.lua (hunk 12 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: How about the second point?

File: src/box/lua/config/instance_config.lua [hunk 16]
```diff
@@ -229,7 +229,30 @@ local function uri_is_suitable_to_connect(uri)
     return true
 end
 
-local function advertise_peer_uri_validate(data, w)
+local function validate_uri_str(data, w)
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:232]: I think we can do some simplifications. Please, check out my proposal below.

```diff
diff --git a/src/box/lua/config/instance_config.lua b/src/box/lua/config/instance_config.lua
index 570be8b3ac..0c742a9025 100644
--- a/src/box/lua/config/instance_config.lua
+++ b/src/box/lua/config/instance_config.lua
@@ -229,60 +229,39 @@ local function uri_is_suitable_to_connect(uri)
     return true
 end
 
-local function validate_uri_str(data, w)
-    --
-    -- Substitute variables with placeholders to don't confuse the URI parser
-    -- with the curly brackets.
-    --
-    data = data:gsub('{{ *.- *}}', 'placeholder')
-    local uri, err = urilib.parse(data)
-    if uri == nil then
-        w.error('Unable to parse an URI: %s', err)
-    end
-    if uri.login ~= nil then
-        w.error("Login must be set via the 'login' option")
-    end
-    assert(uri.password == nil)
-    if uri.params ~= nil then
-        w.error("Parameters must be set via the 'params' option")
-    end
-    local ok, err = uri_is_suitable_to_connect(uri)
-    if not ok then
-        w.error(err)
-    end
-end
-
-local function advertise_client_uri_validate(data, w)
-    -- Accept the special syntax user@ or user:pass@.
-    --
-    -- It means using iproto.listen value to connect and using the
-    -- given user and password.
-    --
-    -- If the password is not given, it is extracted from the
-    -- `credentials` section of the config.
-    if data:endswith('@') then
-        return
-    end
-
-    -- Substitute variables with placeholders to don't confuse the
-    -- URI parser with the curly brackets.
-    data = data:gsub('{{ *.- *}}', 'placeholder')
+-- Verify 'uri' field in iproto.listen and iproto.advertise.*
+-- options.
+local function validate_uri_field(has_login_field, used_to_connect)
+    return function(data, w)
+        -- Substitute variables with placeholders to don't confuse
+        -- the URI parser with the curly brackets.
+        data = data:gsub('{{ *.- *}}', 'placeholder')
 
-    local uri, err = urilib.parse(data)
-    if uri == nil then
-        if data:find(',') then
-            w.error('A single URI is expected, not a list of URIs')
+        local uri, err = urilib.parse(data)
+        if uri == nil then
+            if data:find(',') then
+                w.error('A single URI is expected, not a list of URIs')
+            end
+            w.error('Unable to parse an URI: %s', err)
+        end
+        if uri.login ~= nil and has_login_field then
+            w.error("Login must be set via the 'login' option")
+        end
+        if uri.login ~= nil and not has_login_field then
+            w.error('Login cannot be set for as part of the URI')
+        end
+        assert(uri.password == nil)
+        if uri.params ~= nil then
+            w.error("URI parameters should be described in the 'params' " ..
+                "field, not as the part of URI")
+        end
+        if used_to_connect then
+            local ok, err = uri_is_suitable_to_connect(uri)
+            if not ok then
+                w.error(err)
+            end
         end
-        w.error('Unable to parse an URI: %s', err)
-    end
-    local ok, err = uri_is_suitable_to_connect(uri)
-    if not ok then
-        w.error(err)
     end
-
-    -- The return value is ignored by schema.lua, but useful in
-    -- iproto.advertise.client validation.
-    return uri
 end
 
 -- Accept a value of 'iproto.listen' option and return the first URI
@@ -300,6 +279,10 @@ local function find_suitable_uri_to_connect(listen)
             -- argument `write_password`. Let's assume that the
             -- given URIs are to listen on them and so have no
             -- user/password.
+            --
+            -- NB: We need to format the URI back to construct the
+            -- 'uri' field. urilib.parse() creates separate 'host'
+            -- and 'service'.
             return urilib.format(uri), u.params
         end
     end
@@ -329,6 +312,9 @@ end
 
 local function instance_uri(self, iconfig, advertise_type)
     assert(advertise_type == 'peer' or advertise_type == 'sharding')
+
+    -- An effective value of iproto.advertise.sharding defaults to
+    -- iproto.advertise.peer.
     local uri
     if advertise_type == 'sharding' then
         uri = self:get(iconfig, 'iproto.advertise.sharding')
@@ -339,6 +325,8 @@ local function instance_uri(self, iconfig, advertise_type)
     if advertise_type == 'peer' then
         uri = self:get(iconfig, 'iproto.advertise.peer')
     end
+
+    -- If there is no uri field, look it in iproto.listen.
     uri = uri ~= nil and uri or {}
     if uri.uri == nil then
         assert(uri.params == nil)
@@ -348,12 +336,18 @@ local function instance_uri(self, iconfig, advertise_type)
         end
         uri.uri, uri.params = find_suitable_uri_to_connect(listen)
     end
+
+    -- No URI found for the given instance.
     if uri.uri == nil then
         return nil
     end
+
+    -- If there is a login. but there are no password: lookup the
+    -- credentials section.
     if uri.password == nil and uri.login ~= nil then
         uri.password = find_password(self, iconfig, uri.login)
     end
+
     return uri
 end
 
@@ -744,24 +738,7 @@ return schema.new('instance_config', schema.record({
             items = schema.record({
                 uri = schema.scalar({
                     type = 'string',
-                    validate = function(data, w)
-                        -- Substitute variables with placeholders to don't
-                        -- confuse the URI parser with the curly brackets.
-                        data = data:gsub('{{ *.- *}}', 'placeholder')
-                        local uri, err = urilib.parse(data)
-                        if uri == nil then
-                            w.error('Unable to parse an URI: %s', err)
-                        end
-                        if uri.login ~= nil then
-                            w.error("Login cannot be set for iproto.listen")
-                        end
-                        assert(uri.password == nil)
-                        if uri.params ~= nil then
-                            err = "URI parameters should be described in " ..
-                                  "the 'params' field, not as the part of URI"
-                            w.error(err)
-                        end
-                    end,
+                    validate = validate_uri_field(false, false),
                 }),
                 params = schema.record({
                     transport = schema.enum({
@@ -791,9 +768,8 @@ return schema.new('instance_config', schema.record({
                 }),
             }, {
                 validate = function(data, w)
-                    -- No items in iproto.listen at all -- OK.
                     if next(data) == nil then
-                        return
+                        w.error('An URI should have at least one field')
                     end
                     -- There is some data -- the URI should be there.
                     if data.uri == nil then
@@ -835,48 +811,38 @@ return schema.new('instance_config', schema.record({
         --
         -- The iproto.advertise.{peer,sharding} options have the
         -- following syntax variants:
+        --
+        -- With URI:
+        --
         -- {uri = ...}
         -- {uri = ..., params = ...}
-        -- {login = ...}
-        -- {login = ..., password = ...}
         -- {login = ..., uri = ...}
         -- {login = ..., password = ..., uri = ...}
         -- {login = ..., uri = ..., params = ...}
         -- {login = ..., password = ..., uri = ..., params = ...}
         --
+        -- Without URI:
+        --
+        -- {login = ...}
+        -- {login = ..., password = ...}
+        --
         -- where uri is host:port.
         --
         -- Note: the host:port part may represent a Unix domain
         -- socket: host = 'unix/', port = '/path/to/socket'.
         --
-        -- If there is no host:port (1, 2), it is to be looked in
+        -- If there is no uri field, it is to be looked in
         -- iproto.listen.
         --
-        -- If there is a user, but no password (1, 3), the
-        -- password is to be looked in the `credentials` section
-        -- of the configuration (except user 'guest', which can't
-        -- have a password).
+        -- If there is a user, but no password, the password is to
+        -- be looked in the `credentials` section of the
+        -- configuration (except user 'guest', which can't  have a
+        -- password).
         advertise = schema.record({
             client = schema.scalar({
                 type = 'string',
                 default = box.NULL,
-                validate = function(data, w)
-                    -- Re-use peer URI validation code, but add
-                    -- several extra constraints.
-                    local uri = advertise_client_uri_validate(data, w)
-
-                    if data:endswith('@') then
-                        w.error('user@ and user:pass@ syntax is not ' ..
-                            'accepted by iproto.advertise.client option: ' ..
-                            'only host:port is considered valid')
-                    end
-
-                    if uri.login ~= nil or uri.password ~= nil then
-                        w.error('user@host:port and user:pass@host:port ' ..
-                            'syntax is not accepted by iproto.advertise.' ..
-                            'client option: only host:port is considered valid')
-                    end
-                end,
+                validate = validate_uri_field(false, true),
             }),
             peer = schema.record({
                 login = schema.scalar({
@@ -887,7 +853,7 @@ return schema.new('instance_config', schema.record({
                 }),
                 uri = schema.scalar({
                     type = 'string',
-                    validate = validate_uri_str,
+                    validate = validate_uri_field(true, true),
                 }),
                 params = schema.record({
                     transport = schema.enum({
@@ -915,9 +881,8 @@ return schema.new('instance_config', schema.record({
                 }),
             }, {
                 validate = function(data, w)
-                    -- No data in iproto.advertise.peer -- OK.
                     if next(data) == nil then
-                        return
+                        w.error('An URI should have at least one field')
                     end
                     -- If a password is set, a login must also be specified.
                     if data.password ~= nil and data.login == nil then
@@ -938,7 +903,7 @@ return schema.new('instance_config', schema.record({
                 }),
                 uri = schema.scalar({
                     type = 'string',
-                    validate = validate_uri_str,
+                    validate = validate_uri_field(true, true),
                 }),
                 params = schema.record({
                     transport = schema.enum({
@@ -966,9 +931,8 @@ return schema.new('instance_config', schema.record({
                 }),
             }, {
                 validate = function(data, w)
-                    -- No data in iproto.advertise.sharding -- OK.
                     if next(data) == nil then
-                        return
+                        w.error('An URI should have at least one field')
                     end
                     -- If a password is set, a login must also be specified.
                     if data.password ~= nil and data.login == nil then
```

The list of changes:

1. Reduce code duplication in the validation logic. It also makes eliminates remaining (dead) code that interprets `user@` and `user:pass@` syntax. And give better error messages in come cases (comma separates URIS for example).
2. Forbid `{}` as an URI to simplify logic in `instance_uri`.
3. Refresh the comments.

**Review by Totktonada — APPROVED** [review]
Thank you for all this work! I'm OK in general.

I would like to polish some details a bit more to make the code more clear and consistent. Please, check out my comments.

File: src/box/lua/config/instance_config.lua (hunk 12 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Sorry, I forgot about it. Fixed.

File: src/box/lua/config/configdata.lua (hunk 15 above)
Comment (ImeevMA) [reply src/box/lua/config/configdata.lua]: Added instance name. Also, added a test.

File: src/box/lua/config/instance_config.lua (hunk 16 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua]: Thanks for the suggestion! Applied with some small changes, also divided this diff into two parts - for the first patch (bigger part) and for the second patch (smaller part).

**Comment by Totktonada** [comment]
I looked over the patches once again after the changes and it looks OK. I'm going to push it.

**Comment by Totktonada** [comment]
I've added a temporary workaround for a GC finalizer test.

```diff
commit aa91dcd6c905e842e534eb5d41228970ae530194
Author: Alexander Turenko <alexander.turenko@tarantool.org>
Date:   Wed Dec 20 02:23:54 2023 +0300

    ci: temporarily patch GC finalizer test
    
    It fails in a quite stable manner after the previous commits. Let's
    workaround it and fix it in the LuaJIT submodule later.
    
    NO_DOC=no API changes
    NO_CHANGELOG=no behavior changes
    NO_TEST=nothing to test

diff --git a/.test.mk b/.test.mk
index d4bfd1102f..521645a63c 100644
--- a/.test.mk
+++ b/.test.mk
@@ -45,6 +45,12 @@ build: configure
 
 .PHONY: run-luajit-test
 run-luajit-test:
+	# Temporary hack for an unstable LuaJIT test. To be removed in
+	# the scope of gh-9145.
+	sed -e 's/for _ = 1, 4000 do/for _ = 1, 10000 do/' \
+		-i third_party/luajit/test/tarantool-tests/lj-946-print-errors-from-gc-fin-custom.test.lua
+	sed -e 's/for _ = 1, 4000 do/for _ = 1, 10000 do/' \
+		-i third_party/luajit/test/tarantool-tests/lj-946-print-errors-from-gc-fin-default/script.lua
 	${LUAJIT_TEST_ENV} cmake --build ${LUAJIT_TEST_BUILD_DIR} --parallel ${NPROC} --target LuaJIT-test
 
 .PHONY: install-test-deps
```

CC @fckxorg, @igormunkin

**Pull request merged**